src/main/scala/cromiam/samiam/SamSupervisor.scala

```diff
@@ -0,0 +1,47 @@
+package cromiam.samiam
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.actor.{Actor, ActorLogging, ActorRef, Props, SupervisorStrategy, Terminated}
+import cromiam.samiam.SamIamActor.{SamFailed, SamIamActorRequest}
+
+class SamSupervisor(userIdHeader: String, allowedUsers: List[String]) extends Actor with ActorLogging {
+
+  override def supervisorStrategy: SupervisorStrategy = SupervisorStrategy.stoppingStrategy
+
+  private val samCounter: AtomicInteger = new AtomicInteger(0)
+  private var samIamActor: ActorRef = newSamActor()
+  private var requesters: List[ActorRef] = List.empty
+
+  override def receive: Receive = {
+    case req: SamIamActorRequest =>
+      requesters :+= sender
+      context.watch(sender)
+      samIamActor.forward(req)
+    case Terminated(s) if s == samIamActor =>
+      requesters foreach { _ ! SamFailed }
+      requesters = List.empty
+      regenerateSamIamActor()
+    case Terminated(requester) =>
+      requesters = requesters diff List(requester)
+    case other =>
+      log.error(s"Unexpected message to SamSupervisor: $other")
+  }
+
+  private def regenerateSamIamActor(): Unit = {
+    log.info("Regenerating SAM actor after unexpected shutdown")
+    samIamActor = newSamActor()
+  }
+
+  private def newSamActor(): ActorRef = {
+    val actor = context.actorOf(SamIamActor.props(userIdHeader, allowedUsers), name = s"SamActor_${samCounter.getAndIncrement()}")
```

Comment:
Basically, yes. `ActorLogging` messages are given the actor's name and it's way easier to tell what's going on with real names rather than the UUIDs they get assigned otherwise. The numbers have to be there because the names have to be unique